### ISSUE TITLE
trampoline causes: "returns a value referencing data owned by the current function"

### ISSUE THREAD

**Issue by bbarker** [alecdotninja/tailcall#10]
When we trampoline, we may run into Rust's borrow checker where we previously hadn't:

```rust
error[E0515]: cannot return value referencing temporary value
   --> src/tabs.rs:137:36
    |
136 |   ...                   let buf_iter = buf_new?.iter();
    |                                        -------- temporary value created here
137 |   ...                   return Ok(tailcall::trampoline::Recurse((
    |  ______________________________^
138 | | ...                       f_out, bytes_iter, buf_iter, 1,
139 | | ...                   )));
    | |_________________________^ returns a value referencing data owned by the current function
```

This is because, when we use Rust's recursion, we aren't actually returning anything from the function in the recursive case, but when we trampoline, we create the inner function (closure) where we do return.

I'm not sure if there's a clever way around this; the best thing might be to delineate some alternative solutions in the README/docs.

**Comment by bbarker** [comment]
Possible solution strategies:

* Iterators
  * Use `.into_iter()` instead of `.iter()` to generate a `std::vec::IntoIter<_>` instead of a  `std::slice::Iter<'_, u8>`.
* (*Generates overhead*) (Need to test): Use `Rc<RefCell>` or `Arc<Mutex>` to (probably only the former generally, as recursion is not parallelelizable) to avoid the borrow checker

Some of these solutions we may eventually be able to build into the macro, though I'm not sure if it is worth it.

**Comment by alecdotninja** [comment]
This may be solved by changing the signature of `trampoline::run`:

```rust
#[inline(always)]
pub fn run<StepFn, Input, Output>(mut step: StepFn, mut input: Input) -> Output
where
    StepFn: FnMut(Input) -> Next<Input, Output>,
{
    loop {
        match step(input) {
            Recurse(new_input) => {
                input = new_input;
                continue;
            }
            Finish(output) => {
                break output;
            }
        }
    }
}
```

It probably should have been this all along since we can statically guarantee the closure is only run "once at a time".

**Comment by alecdotninja** [comment]
This was fixed via #11 which was released in [v0.1.5](https://crates.io/crates/tailcall/0.1.5). Thank you for all of your help here, @bbarker !
